[llvm/Support] Make `llvm::sys::RWMutex` Lockable

medismailben opened the pull request:
This patch extends the `llvm::sys::RWMutex` class to fullfill the `Lockable` requirement to include attempted locking, by implementing a `bool try_lock` member function.

As the name suggests, this method will try to acquire to lock in a non-blocking fashion and release it immediately. If it managed to acquire the lock, returns `true`, otherwise returns `false`.

bulbazord commented on llvm/lib/Support/RWMutex.cpp:
```diff
@@ -107,6 +108,14 @@ RWMutexImpl::unlock()
+  pthread_rwlock_t *rwlock = static_cast<pthread_rwlock_t *>(data_);
+  assert(rwlock != nullptr);
+
+  int errorcode = pthread_rwlock_tryrdlock(rwlock);
```
There's an inconsistency here. `RWMutexImpl` has `shared` variants of `lock` and `unlock` that are used for reading while the plain versions of `lock` and `unlock` are for writing. I think this needs to use `pthread_rwlock_trywrlock` and then you probably want to add `RWMutexImpl::try_lock_shared()`.

medismailben committed:
[llvm/Support] Make `llvm::sys::RWMutex` Lockable

This patch extends the `llvm::sys::RWMutex` class to fullfill the
`Lockable` requirement to include attempted locking, by implementing a
`bool try_lock` member function.

As the name suggests, this method will try to acquire to lock in a
non-blocking fashion and release it immediately. If it managed to
acquire the lock, returns `true`, otherwise returns `false`.

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>

bulbazord reviewed (approved):
LGTM

JDevlieghere reviewed (changes requested):
It looks like this is trying to implement something like `is_locked`. `try_lock` has fairly well defined semantics for `std::mutex` and `pthreads`. In both cases this acquires the lock if it succeeds. Our implementation shouldn't behave differently.

JDevlieghere reviewed (approved):
This LGTM!

dwblaikie commented:
as an aside - per the instructions here ( https://llvm.org/docs/GitHub.html#updating-pull-requests ) please don't force-push to pull request reviews, as it makes it difficult to trace back comments and discussion through the revisions of a patch (I think, especially, inline comments can get lost/hidden/hard to find when force pushing)

medismailben commented:
> as an aside - per the instructions here ( https://llvm.org/docs/GitHub.html#updating-pull-requests ) please don't force-push to pull request reviews, as it makes it difficult to trace back comments and discussion through the revisions of a patch (I think, especially, inline comments can get lost/hidden/hard to find when force pushing)

Sure! I'll keep that in mind next time.
